File: trion/C/misc/ListBoards/list_boards.c
/**
 * Short example to showing how to list all installed boards.
 * Shows: - used configuration directory
 *        - BoardID, Board Name, SlotNo
 *
 * Note: look into the generated Board[0..127]_Properties.xml to see were
 * SlotNo and SerialNo information is stored.
 */

#include "dewepxi_apicore.h"
#include "dewepxi_apiutil.h"
#include "dewepxi_types.h"
#include "trion_sdk_util.h"
#include <stdio.h>

int doListBoards();

int nUseAllCommands = 0;
int nUseBoardReset = 0;
int nUseThreads = 0;
int nUseMergeCache = 0;
int nShowKey = 0;
int ntestLoop = 1;

int main(int argc, char* argv[])
{
    int nErrorCode = 0;
    const char* sThreadPoolSize = 0;
    int i = 0;

    if (argc > 1)
    {
        for (i = 1; i < argc; ++i)
        {
            if (strncmp(argv[i], "--all", sizeof("--all")) == 0)
            {
                // use advanced board reset
                nUseAllCommands = 1;
            }

            if (strncmp(argv[i], "--reset", sizeof("--reset")) == 0)
            {
                // use advanced board reset
                nUseBoardReset = 1;
            }
            else if (strncmp(argv[i], "--thread", sizeof("--thread")) == 0)
            {
                // enable the usage of thread
                nUseThreads = 1;
            }
            else if (strncmp(argv[i], "--pool", sizeof("--pool")) == 0)
            {
                if (argc > (i + 1))
                {
                    sThreadPoolSize = argv[i + 1];
                }
            }
            else if (strncmp(argv[i], "--runs", sizeof("--runs")) == 0)
            {
                if (argc > (i + 1))
                {
                    ntestLoop = atoi(argv[i + 1]);
                }
            }
            else if (strncmp(argv[i], "--cache", sizeof("--cache")) == 0)
            {
                // enable the usage of thread
                nUseMergeCache = 1;
            }
            else if (strncmp(argv[i], "--key", sizeof("--key")) == 0)
            {
                // enable the usage of thread
                nShowKey = 1;
            }
            else if (strncmp(argv[i], "--help", sizeof("--help")) == 0)
            {
                printf("Usage: Listboards [-all] [--reset] [--thread] [--pool NUM] [--cache]\n");
                printf("  --all         use _ALL commands for open,close,reset\n");
                printf("  --reset       do a board reset\n");
                printf("  --thread      use threads in the TRION API for performance improvements\n");
                printf("  --pool NUM    used number of threads in the TRION API for performance improvements\n");
                printf("  --runs NUM    number of re-runs of this example\n");
                printf("  --cache       use cached merge documents\n");
                printf("  --key         show board key info\n");
                printf("  --help    s   how this usage screen\n");
                return 0;
            }
        }

    }

    while (--ntestLoop >= 0)
    {
        printf("Test Loop: %d\n\n", ntestLoop);

        // Load pxi_api.dll
        if (0 != LoadTrionApi())
        {
            return 1;
        }


        if (nUseThreads)
        {
            char buffer[1024] = {0};

            nErrorCode = DeWeSetParamStruct_str("driver/api/config/thread", "Enabled", "true");
            CheckError(nErrorCode);

            if (sThreadPoolSize)
            {
                nErrorCode = DeWeSetParamStruct_str("driver/api/config/thread", "PoolSize", sThreadPoolSize);
                CheckError(nErrorCode);
            }

        }

        if (nUseMergeCache)
        {
            nErrorCode = DeWeSetParamStruct_str("driver/api/config/xml", "AllowCachedMergeResult", "true");
            CheckError(nErrorCode);
        }
        else
        {
            nErrorCode = DeWeSetParamStruct_str("driver/api/config/xml", "AllowCachedMergeResult", "false");
            CheckError(nErrorCode);
        }

        printf("Options used:\n");
        printf("  --all    => %s\n", nUseAllCommands ? "yes" : "no");
        printf("  --reset  => %s\n", nUseBoardReset ? "yes" : "no");
        printf("  --thread => %s\n", nUseThreads ? "yes" : "no");
        printf("  --pool   => %s\n", sThreadPoolSize);
        printf("  --cache  => %s\n", nUseMergeCache ? "yes" : "no");

        // use sim:
        DeWeSetParamStruct_str("driver/api/config/simulation", "mode", "SIMMODE_FORCE");
        doListBoards();

        // use hw:
        DeWeSetParamStruct_str("driver/api/config/simulation", "mode", "SIMMODE_OFF");
        doListBoards();



        // Unload pxi_api.dll
        UnloadTrionApi("\nEnd Of Example\n");
    }

    return nErrorCode;
}


int doListBoards()
{
    int nNoOfBoards = 0;
    int nErrorCode = 0;
    int nBoardID = 0;
    int nSlotNo = 0;
    int nSegmentNo = 0; // for internal chassis controller
    int nEncID = 0;
    char sPath[256]      = { 0 };
    char sBoardName[256] = {0};
    char sSerialNo[256] = {0};
    char sEncID[256]    = {0};
    char sSlotNo[256]   = {0};
    char sSegmentNo[256]  = {0};
    char sBoardID[256]  = {0};
    char sKey[1024] = { 0 };
    char sBuffer[2048] = { 0 };

    // Initialize driver and retrieve the number of TRION boards
    // nNoOfBoards is a negative number if system is in DEMO mode!
    nErrorCode = DeWeDriverInit(&nNoOfBoards);
    CheckError(nErrorCode);

    if (nNoOfBoards == 0)
    {
        printf("No TRION board found\n");
    }

    if (nNoOfBoards < 0)
    {
        printf("Trion API is set to use simulated boards\n");
    }
    else
    {
        printf("Trion API is set to use real boards\n");
    }

    nNoOfBoards = abs(nNoOfBoards);

    if (nNoOfBoards == 0)
    {
        printf("No boards found\n");
        DeWeDriverDeInit();
        return ERR_NONE;
    }


    nErrorCode = DeWeGetParamStruct_str("driver/api", "SystemVersions", sBuffer, sizeof(sBuffer));
    //if (ERR_NONE == nErrorCode) printf("Trion Versions : %s\n", sBuffer);

    nErrorCode = GetApiPath(TRION_CONFIG_PATH, sPath, sizeof(sPath));
    if (ERR_NONE == nErrorCode) printf("Trion API config path : %s\n", sPath);

    nErrorCode = GetApiPath(TRION_LOG_PATH, sPath, sizeof(sPath));
    if (ERR_NONE == nErrorCode) printf("Trion API log path    : %s\n", sPath);

    nErrorCode = GetApiPath(TRION_BACKUP_PATH, sPath, sizeof(sPath));
    if (ERR_NONE == nErrorCode) printf("Trion API backup path : %s\n", sPath);

    // enclosure info
    {
        char sEncName[32] = {0};
        snprintf(sBoardID, sizeof(sBoardID), "BoardID%d/boardproperties/SystemInfo/EnclosureInfo", 0);
        nErrorCode = DeWeGetParamXML_str(sBoardID, "Name", sEncName, sizeof(sEncName));
        CheckError(nErrorCode); 
        printf("Enclosure name: %s\n", sEncName);
    }

    printf("%-7s %-7s %-8s %-30s %s\n", "EncID", "SlotNo[i]", "BoardID", "Name", "SerialNo");
    printf("------------------------------------------------------------------------\n");

    // iterate all boards for open
    if (nUseAllCommands)
    {
        // Open ALL
        nErrorCode = DeWeSetParam_i32(0, CMD_OPEN_BOARD_ALL, 0);
        CheckError(nErrorCode);
    }
    else
    {
        for (nBoardID = 0; nBoardID < nNoOfBoards; ++nBoardID)
        {
            // Open
            nErrorCode = DeWeSetParam_i32(nBoardID, CMD_OPEN_BOARD, 0);
            CheckError(nErrorCode);
        }
    }

    // iterate all boards for reset
    if (nUseBoardReset)
    {
        if (nUseAllCommands)
        {
            // Reset ALL
            nErrorCode = DeWeSetParam_i32(0, CMD_RESET_BOARD_ALL, 0);
            CheckError(nErrorCode);
        }
        else
        {
            for (nBoardID = 0; nBoardID < nNoOfBoards; ++nBoardID)
            {
                // Reset
                nErrorCode = DeWeSetParam_i32(nBoardID, CMD_RESET_BOARD, 0);
                CheckError(nErrorCode);
            }
        }
    }


    // iterate all boards
    for (nBoardID = 0; nBoardID < nNoOfBoards; ++nBoardID)
    {
        // Get the Enclosure ID:
        // Path to <SystemInfo> element in Board[0..127]Properties.xml
        snprintf(sBoardID, sizeof(sBoardID), "BoardID%d/boardproperties/SystemInfo/EnclosureInfo", nBoardID);
        nErrorCode = DeWeGetParamXML_str(sBoardID, "EnclosureID", sEncID, sizeof(sEncID));
        CheckError(nErrorCode);
        sscanf(sEncID, "%d", &nEncID);

        // Get the SlotNo:
        // Path to <SystemInfo> element in Board[0..127]Properties.xml
        //   -> BoardID%d/SystemInfo  (root element <Properies> is ignored by API)
        snprintf(sBoardID, sizeof(sBoardID), "BoardID%d/boardproperties/SystemInfo", nBoardID);
        nErrorCode = DeWeGetParamXML_str(sBoardID, "SlotNo", sSlotNo, sizeof(sSlotNo));
        CheckError(nErrorCode);
        sscanf(sSlotNo, "%d", &nSlotNo);

        // Slot ID for internal controller devices (DEWE3)
        nErrorCode = DeWeGetParamXML_str(sBoardID, "InternalSegmentNo", sSegmentNo, sizeof(sSegmentNo));
        CheckError(nErrorCode);
        sscanf(sSegmentNo, "%d", &nSegmentNo);

        // Build a string in the format: "BoardID0", "BoardID1", ...
        snprintf(sBoardID, sizeof(sBoardID), "BoardID%d", nBoardID);

        // Request the TRION board name
        nErrorCode = DeWeGetParamStruct_str(sBoardID, "BoardName", sBoardName, sizeof(sBoardName));
        CheckError(nErrorCode);

        //Build a string to query the XML-stored serial number
        snprintf(sBoardID, sizeof(sBoardID), "BoardID%d/boardproperties/BoardInfo", nBoardID);
        nErrorCode = DeWeGetParamXML_str(sBoardID, "SerialNumber", sSerialNo, sizeof(sSerialNo));
        CheckError(nErrorCode);

        // Print Properties
        if (nShowKey)
        {
            snprintf(sBoardID, sizeof(sBoardID), "BoardID%d", nBoardID);
            nErrorCode = DeWeGetParamStruct_str(sBoardID, "key", sKey, sizeof(sKey));
            CheckError(nErrorCode);
            printf("  %-7d %-2d[%d] %5d      %-30s %-12s <%s>\n", nEncID, nSlotNo, nSegmentNo, nBoardID, sBoardName, sSerialNo, sKey);
        }
        else
        {
            printf("  %-7d %-2d[%d] %5d      %-30s %-12s\n", nEncID, nSlotNo, nSegmentNo, nBoardID, sBoardName, sSerialNo);
        }
    }


    if (nUseAllCommands)
    {
        // Reset ALL
        nErrorCode = DeWeSetParam_i32(0, CMD_CLOSE_BOARD_ALL, 0);
        CheckError(nErrorCode);
    }
    else
    {
        // iterate all boards
        for (nBoardID = 0; nBoardID < nNoOfBoards; ++nBoardID)
        {
            // Close the board connection
            nErrorCode = DeWeSetParam_i32(nBoardID, CMD_CLOSE_BOARD, 0);
            CheckError(nErrorCode);
        }
    }


    DeWeDriverDeInit();

    return ERR_NONE;
}

File: trion/CS/TRION_SDK_CS/ListBoards/list_boards.cs
using System;
using Trion;

class ListBoardsOptions
{
    public int NumberOfBoards { get; set; } = 0;
    public bool UseAllCommands { get; set; }
    public bool UseBoardReset { get; set; }
    public bool UseThreads { get; set; }
    public bool UseMergeCache { get; set; }
    public bool ShowKey { get; set; }
    public int TestLoop { get; set; } = 1;
    public string ThreadPoolSize { get; set; }
}

class ListBoards
{
    static ListBoardsOptions ParseArguments(string[] args)
    {
        var opts = new ListBoardsOptions();
        for (int i = 0; i < args.Length; ++i)
        {
            switch (args[i])
            {
                case "--all": opts.UseAllCommands = true; break;
                case "--reset": opts.UseBoardReset = true; break;
                case "--thread": opts.UseThreads = true; break;
                case "--cache": opts.UseMergeCache = true; break;
                case "--key": opts.ShowKey = true; break;
                case "--pool":
                    if (i + 1 < args.Length)
                    {
                        opts.ThreadPoolSize = args[i + 1];
                    }
                    break;
                case "--runs":
                    if (i + 1 < args.Length && int.TryParse(args[i + 1], out int runs))
                    {
                        opts.TestLoop = runs;
                    }
                    break;
                case "--help":
                    PrintUsage();
                    Environment.Exit(0);
                    break;
            }
        }
        return opts;
    }

    static void PrintUsage()
    {
        Console.WriteLine("Usage: ListBoards [-all] [--reset] [--thread] [--pool NUM] [--cache]");
        Console.WriteLine("  --all         use _ALL commands for open,close,reset");
        Console.WriteLine("  --reset       do a board reset");
        Console.WriteLine("  --thread      use threads in the TRION API for performance improvements");
        Console.WriteLine("  --pool NUM    used number of threads in the TRION API for performance improvements");
        Console.WriteLine("  --runs NUM    number of re-runs of this example");
        Console.WriteLine("  --cache       use cached merge documents");
        Console.WriteLine("  --key         show board key info");
        Console.WriteLine("  --help        show this usage screen");
    }

    static void PrintOptionsUsed(ListBoardsOptions opts)
    {
        Console.WriteLine("Options used:");
        Console.WriteLine("  --all    => {0}", opts.UseAllCommands ? "yes" : "no");
        Console.WriteLine("  --reset  => {0}", opts.UseBoardReset ? "yes" : "no");
        Console.WriteLine("  --thread => {0}", opts.UseThreads ? "yes" : "no");
        Console.WriteLine("  --pool   => {0}", opts.ThreadPoolSize ?? "(null)");
        Console.WriteLine("  --cache  => {0}", opts.UseMergeCache ? "yes" : "no");
    }

    static void DoListBoards(ListBoardsOptions opts)
    {
        var current_number_of_boards = opts.NumberOfBoards;
        Trion.TrionError error;
        if (opts.NumberOfBoards == 0)
        {
            Console.WriteLine("No TRION board found");
            TrionApi.Uninitialize();
            Environment.Exit(0);
        }
        if (opts.NumberOfBoards > 0)
        {
            Console.WriteLine("Trion API is set to use real boards");
        }
        else
        {
            Console.WriteLine("Trion API is set to use simulated boards");
        }

        current_number_of_boards = Math.Abs(current_number_of_boards);

        // get the config path
        (error, string config_path) = TrionApi.DeWeGetParamStruct_String("System/config", "Path");
        if (error != TrionError.NONE || string.IsNullOrEmpty(config_path))
        {
            Console.WriteLine("Error retrieving config path: " + Trion.API.DeWeErrorConstantToString(error));
        }
        else
        {
            Console.WriteLine("Trion API config path       : " + config_path);
        }

        // get the log path
        (error, string log_path) = TrionApi.DeWeGetParamStruct_String("System/log", "Path");
        if (error != TrionError.NONE || string.IsNullOrEmpty(log_path))
        {
            Console.WriteLine("Error retrieving log path : " + Trion.API.DeWeErrorConstantToString(error));
        }
        else
        {
            Console.WriteLine("Trion API log path          : " + log_path);
        }

        // get the api backup path
        (error, string backup_path) = TrionApi.DeWeGetParamStruct_String("System/backup", "Path");
        if (error != TrionError.NONE || string.IsNullOrEmpty(backup_path))
        {
            Console.WriteLine("Error retrieving backup path: " + Trion.API.DeWeErrorConstantToString(error));
        }
        else
        {
            Console.WriteLine("Trion API backup path       : " + backup_path);
        }

        // get enclosure name
        (error, string enclosure_name) = TrionApi.DeWeGetParamXML_String("BoardID0/boardproperties/SystemInfo/EnclosureInfo", "Name");
        if (error != TrionError.NONE || string.IsNullOrEmpty(enclosure_name))
        {
            Console.WriteLine("Error retrieving enclosure name: " + Trion.API.DeWeErrorConstantToString(error));
        }
        else
        {
            Console.WriteLine("Enclosure name: " + enclosure_name);
        }

        Console.WriteLine("{0,-8} {1,-7} {2,-7} {3,-30} {4}", "EncID", "SlotNo", "BoardID", "Name", "SerialNo");
        Console.WriteLine("--------------------------------------------------------------------------");

        if (opts.UseAllCommands)
        {
            error = TrionApi.DeWeSetParam_i32(0, Trion.TrionCommand.OPEN_BOARD_ALL, 0);
            if (error != TrionError.NONE)
            {
                Console.WriteLine("Failed to open all boards: " + Trion.API.DeWeErrorConstantToString(error));
                return;
            }
        }
        else
        {
            for (int current_board_id = 0; current_board_id < current_number_of_boards; ++current_board_id)
            {
                error = TrionApi.DeWeSetParam_i32(current_board_id, Trion.TrionCommand.OPEN_BOARD, 0);
                if (error != TrionError.NONE)
                {
                    Console.WriteLine($"Failed to open board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                    continue;
                }
            }
        }

        // iterate all boards for reset
        if (opts.UseBoardReset)
        {
            if (opts.UseAllCommands)
            {
                error = TrionApi.DeWeSetParam_i32(0, Trion.TrionCommand.RESET_BOARD_ALL, 0);
                if (error != TrionError.NONE)
                {
                    Console.WriteLine("Failed to reset all boards: " + Trion.API.DeWeErrorConstantToString(error));
                    return;
                }
            }
            else
            {
                for (int current_board_id = 0; current_board_id < current_number_of_boards; ++current_board_id)
                {
                    error = TrionApi.DeWeSetParam_i32(current_board_id, Trion.TrionCommand.RESET_BOARD, 0);
                    if (error != TrionError.NONE)
                    {
                        Console.WriteLine($"Failed to reset board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                        continue;
                    }
                }
            }
        }

        // iterate all boards for info
        for (int current_board_id = 0; current_board_id < current_number_of_boards; ++current_board_id)
        {
            // get the enclosure ID:
            // Path to <SystemInfo> element in Board[0..127]Properties.xml
            (error, string enclosure_id) = TrionApi.DeWeGetParamXML_String($"BoardID{current_board_id}/boardproperties/SystemInfo/EnclosureInfo", "EnclosureID");
            if (error != TrionError.NONE || string.IsNullOrEmpty(enclosure_id))
            {
                Console.WriteLine($"Failed to get enclosure ID for board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                continue;
            }

            // get the slot number
            // Path to <SystemInfo> element in Board[0..127]Properties.xml
            //   -> BoardID%d/SystemInfo  (root element <Properies> is ignored by API)
            (error, string slot_no) = TrionApi.DeWeGetParamXML_String($"BoardID{current_board_id}/boardproperties/SystemInfo", "SlotNo");
            if (error != TrionError.NONE || string.IsNullOrEmpty(slot_no))
            {
                Console.WriteLine($"Failed to get slot number for board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                continue;
            }

            // slot id for internal controller device (DEWE3)
            (error, string segment_no) = TrionApi.DeWeGetParamXML_String($"BoardID{current_board_id}/boardproperties/SystemInfo", "InternalSegmentNo");
            if (error != TrionError.NONE || string.IsNullOrEmpty(segment_no))
            {
                Console.WriteLine($"Failed to get internal segment number for board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                continue;
            }

            // request the TRION board name
            (error, string board_name) = TrionApi.DeWeGetParamStruct_String($"BoardID{current_board_id}", "BoardName");
            if (error != TrionError.NONE || string.IsNullOrEmpty(board_name))
            {
                Console.WriteLine($"Failed to get board name for board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                continue;
            }

            // get the board serial number (for simulated boards this is always 12345678)
            (error, string serial_no) = TrionApi.DeWeGetParamXML_String($"BoardID{current_board_id}/boardproperties/BoardInfo", "SerialNumber");
            if (error != TrionError.NONE || string.IsNullOrEmpty(serial_no))
            {
                Console.WriteLine($"Failed to get serial number for board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                continue;
            }

            // print properties
            if (opts.ShowKey)
            {
                (error, string key) = TrionApi.DeWeGetParamStruct_String($"BoardID{current_board_id}", "key");
                if (error != TrionError.NONE || string.IsNullOrEmpty(key))
                {
                    Console.WriteLine($"Failed to get key for board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                    continue;
                }
                Console.WriteLine($"  {enclosure_id,-7} {slot_no,-2}[{segment_no}] {current_board_id,5}      {board_name,-30} {serial_no,-12} <{key}>");
            }
            else
            {
                Console.WriteLine($"  {enclosure_id,-7} {slot_no,-2}[{segment_no}] {current_board_id,5}      {board_name,-30} {serial_no,-12}");
            }
        }
    }

    static void Main(string[] args)
    {
        var opts = ParseArguments(args);
        for (int loop = 0; loop < opts.TestLoop; ++loop)
        {
            Console.WriteLine($"Test Loop: {loop}\n");
            opts.NumberOfBoards = TrionApi.Initialize();
            if (opts.NumberOfBoards == 0)
            {
                Console.WriteLine("No TRION board found");
                continue;
            }

            // enable threading if requested --------------------------------------------------------------------------
            if (opts.UseThreads)
            {
                var errorCode = TrionApi.DeWeSetParamStruct("driver/api/config/thread", "Enabled", "true");
                if (errorCode != TrionError.NONE)
                {
                    Console.WriteLine("Failed to enable threading: " + Trion.API.DeWeErrorConstantToString(errorCode));
                }
                if (!string.IsNullOrEmpty(opts.ThreadPoolSize))
                {
                    errorCode = TrionApi.DeWeSetParamStruct("driver/api/config/thread", "PoolSize", opts.ThreadPoolSize);
                    if (errorCode != TrionError.NONE)
                    {
                        Console.WriteLine("Failed to set thread pool size: " + Trion.API.DeWeErrorConstantToString(errorCode));
                    }
                }
            }

            // enable merge cache if requested ------------------------------------------------------------------------
            if (opts.UseMergeCache)
            {
                var errorCode = TrionApi.DeWeSetParamStruct("driver/api/config/xml", "AllowCachedMergeResult", "true");
                if (errorCode != TrionError.NONE)
                {
                    Console.WriteLine("Failed to enable merge cache: " + Trion.API.DeWeErrorConstantToString(errorCode));
                }
            }
            else
            {
                var errorCode = TrionApi.DeWeSetParamStruct("driver/api/config/xml", "AllowCachedMergeResult", "false");
                if (errorCode != TrionError.NONE)
                {
                    Console.WriteLine("Failed to disable merge cache: " + Trion.API.DeWeErrorConstantToString(errorCode));
                }
            }

            PrintOptionsUsed(opts);
            DoListBoards(opts);

            if (opts.UseAllCommands)
            {
                var error = TrionApi.DeWeSetParam_i32(0, Trion.TrionCommand.CLOSE_BOARD_ALL, 0);
                if (error != TrionError.NONE)
                {
                    Console.WriteLine("Failed to close all boards: " + Trion.API.DeWeErrorConstantToString(error));
                }
            }
            else
            {
                for (int current_board_id = 0; current_board_id < opts.NumberOfBoards; ++current_board_id)
                {
                    var error = TrionApi.DeWeSetParam_i32(current_board_id, Trion.TrionCommand.CLOSE_BOARD, 0);
                    if (error != TrionError.NONE)
                    {
                        Console.WriteLine($"Failed to close board {current_board_id}: " + Trion.API.DeWeErrorConstantToString(error));
                    }
                }
            }

            TrionApi.Uninitialize();
        }
    }
}